### PR TITLE
Allow specify wall directions

### PR DESCRIPTION
**Pull request by Noisyfox**
Add an option "wall direction", which allows you to change the wall direction per-part/object/layer.
Might be useful if you have parts that printed upside down and you want a unified surface appearance when assembled.

### CODE REVIEW

File: src/libslic3r/Layer.cpp [hunk 1]
```diff
@@ -176,6 +176,7 @@ void Layer::make_perimeters()
 		                && config.detect_overhang_wall                   == other_config.detect_overhang_wall
 		                && config.overhang_reverse                       == other_config.overhang_reverse
 		                && config.overhang_reverse_threshold             == other_config.overhang_reverse_threshold
+                        && config.wall_direction                         == other_config.wall_direction
```
Comment (discip) [inline src/libslic3r/Layer.cpp:179]:
```suggestion
                                && config.wall_direction                         == other_config.wall_direction
```

**Comment by Noisyfox** [comment]
This needs to be tested because I was concerned about the effectiveness of reversing the extrusion direction on upside down parts, because the nozzle tip is asymmetrical, so the wall might not look the same when extruded in counter-clockwise vs extruded in clockwise then flip over.

**Comment by kinedix** [comment]
Feature works amazing! exactly as expected, huge kudos!

**Comment by discip** [comment]
> Feature works amazing! exactly as expected, huge kudos! Win app works great, Mac both x86 and Arm, showed as damaged, for some reason.

Do you by any chance have a picture of the result?

**Comment by kinedix** [comment]
@discip yes I do, here is before, and after
before
after

before, bulges was on opposite sides, now they are continuous.
blue line represents connection gap

**Comment by kinedix** [comment]
Actually this helped for Mac. version "xattr -cr"

File: src/libslic3r/PrintConfig.cpp [hunk 2]
```diff
@@ -189,6 +189,14 @@ static t_config_enum_values s_keys_map_WallSequence {
 };
 CONFIG_OPTION_ENUM_DEFINE_STATIC_MAPS(WallSequence)
 
+//Orca
+static t_config_enum_values s_keys_map_WallDirection{
+    { "default", int(WallDirection::Default) },
```
Comment (SoftFever) [inline src/libslic3r/PrintConfig.cpp:194]: I suggest we use 'auto' for clearer clarification.
What do you think?

Comment (Noisyfox) [reply src/libslic3r/PrintConfig.cpp]: Good idea. Done.

**Review by SoftFever — APPROVED** [review]

**Comment by vovodroid** [comment]
Hi, what do you think about not reversing walls here

```
        const WallDirection wall_direction = config->wall_direction;
        if (wall_direction != WallDirection::Auto) {
            // Skip steep overhang detection if wall direction is specified
            steep_overhang_contour = true;
            steep_overhang_hole    = true;
        }
        if (ExtrusionEntityCollection extrusion_coll = traverse_extrusions(*this, ordered_extrusions, steep_overhang_contour, steep_overhang_hole); !extrusion_coll.empty()) {
            // All walls are counter-clockwise initially, so we don't need to reorient it if that's what we want
            if (wall_direction != WallDirection::CounterClockwise) {
                reorient_perimeters(extrusion_coll, steep_overhang_contour, steep_overhang_hole,
                                    // Reverse internal only if the wall direction is auto
                                    this->config->overhang_reverse_internal_only && wall_direction == WallDirection::Auto);
            }
            this->loops->append(extrusion_coll);
        }
```

but creating in desired direction from the vary beginning?

```
            ExtrusionLoop *eloop = static_cast<ExtrusionLoop*>(coll.entities[idx.first]);
            coll.entities[idx.first] = nullptr;

            eloop->make_counter_clockwise();

....
                ExtrusionLoop extrusion_loop(std::move(paths), pg_extrusion.is_contour ? elrDefault : elrHole);
                extrusion_loop.make_counter_clockwise();

```

I enhanced reversing logic and added reversing walls order for overhang, and current logic complicated the issue.

**Comment by vovodroid** [comment]
Also probably no Auto option needed. For forcing direction it's enough just to disable reversing.

**Pull request merged**